src/components/Home.tsx

```diff
@@ -0,0 +1,186 @@
+import React, { useEffect, useState } from 'react';
+import { Link, useNavigate } from 'react-router-dom';
+import { auth, db } from './firebase';
+import { onAuthStateChanged } from 'firebase/auth';
+import { doc, getDoc, setDoc } from 'firebase/firestore';
+import notesImage from './notes.jpg';
+import flashcardsImage from './flashcard.jpg';
+import quizzesImage from './quiz.png';
+import studyGroupsImage from './studygroup.jpg';
+import discussionBoardImage from './discussion.jpg';
+import progressTrackingImage from './progress.png';
+import NavBar from './Navbar';
+import firebase from 'firebase/compat/app';
+import './Home.css';
+interface UserDetails {
+  name: string;
+  age: number;
+  role: string;
+}
+
+const Home: React.FC = () => {
+  const [user, setUser] = useState<firebase.User | null>(null);
+  const [userDetails, setUserDetails] = useState<UserDetails | null>(null);
+  const [needsDetails, setNeedsDetails] = useState<boolean>(false);
+  const navigate = useNavigate();
+
+  useEffect(() => {
+    const unsubscribe = onAuthStateChanged(auth, async (currentUser) => {
+      if (currentUser) {
+        console.log('User logged in:', currentUser);
+        setUser(currentUser);
+        const userDocRef = doc(db, 'users', currentUser.uid);
+        const userDoc = await getDoc(userDocRef);
+        if (!userDoc.exists()) {
+          console.log('User document does not exist. Prompting for details.');
+          setNeedsDetails(true);
+        } else {
+          const userData = userDoc.data() as UserDetails;
+          setUserDetails(userData);
+          setNeedsDetails(false);
+        }
+      } else {
+        console.log('No user logged in. Redirecting to login.');
+        setUser(null);
+        navigate('/login');
+      }
+    });
+
+    return () => unsubscribe();
+  }, [navigate]);
+
+  const handleDetailsSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+    const { name, age, role } = e.currentTarget.elements as typeof e.currentTarget.elements & {
+      name: HTMLInputElement;
+      age: HTMLInputElement;
+      role: RadioNodeList;
+    };
+
+    try {
+      if (!user) throw new Error("User is not logged in");
+      
+      const userData: UserDetails = {
+        name: name.value,
+        age: parseInt(age.value),
+        role: role.value
+      };
+      await setDoc(doc(db, 'users', user.uid), userData);
+      console.log('User details saved successfully.');
+      setUserDetails(userData);
+      setNeedsDetails(false);
+      
+      if (role.value === 'student') {
+        navigate('/student');
+      } else if (role.value === 'teacher') {
+        navigate('/teacher');
+      }
+    } catch (error) {
+      console.error('Error saving user details:', error);
+    }
+  };
+
+  return (
+    <div>
+      <NavBar user={user} userDetails={userDetails} showUserDetails={true} />
+      {needsDetails ? (
```

Comment:
create two different components i.e. one which will run if needsDetails id true and other one when needDetails is false.

Reply:
Also render both of them with useMemo